### ISSUE TITLE
calamari-cross-fold-train fails in make_checkpoint with exception TypeError ("Not JSON Serializable")

### ISSUE THREAD

**Issue by stweil** [Calamari-OCR/calamari#145]
Jetson AGX Xavier is a small unit running Ubuntu Bionic for arm64 and designed for GPU applications. It fails to run `calamari-cross-fold-train` (tested with installation from PyPI and with installation from Git master):
 
    Traceback (most recent call last):
      File "/home/stweil/src/github/qurator-spk/train-calamari-gt4histocr/venv/lib/python3.6/site-packages/calamari_ocr/scripts/train.py", line 371, in <module>
        main()
      File "/home/stweil/src/github/qurator-spk/train-calamari-gt4histocr/venv/lib/python3.6/site-packages/calamari_ocr/scripts/train.py", line 367, in main
        run(args)
      File "/home/stweil/src/github/qurator-spk/train-calamari-gt4histocr/venv/lib/python3.6/site-packages/calamari_ocr/scripts/train.py", line 359, in run
        progress_bar=not args.no_progress_bars
      File "/home/stweil/src/github/qurator-spk/train-calamari-gt4histocr/venv/lib/python3.6/site-packages/calamari_ocr/ocr/trainer.py", line 197, in train
        self._run_train(train_net, train_start_time, progress_bar, self.dataset, self.validation_dataset, training_callback)
      File "/home/stweil/src/github/qurator-spk/train-calamari-gt4histocr/venv/lib/python3.6/site-packages/calamari_ocr/ocr/trainer.py", line 213, in _run_train
        train_net.train(train_dataset, val_dataset, checkpoint_params, self.txt_postproc, progress_bar, training_callback)
      File "/home/stweil/src/github/qurator-spk/train-calamari-gt4histocr/venv/lib/python3.6/site-packages/calamari_ocr/ocr/backends/tensorflow_backend/tensorflow_model.py", line 330, in train
        v_cb, es_cb
      File "/home/stweil/src/github/qurator-spk/train-calamari-gt4histocr/venv/lib/python3.6/site-packages/tensorflow/python/keras/engine/training.py", line 780, in fit
        steps_name='steps_per_epoch')
      File "/home/stweil/src/github/qurator-spk/train-calamari-gt4histocr/venv/lib/python3.6/site-packages/tensorflow/python/keras/engine/training_arrays.py", line 239, in model_iteration
        callbacks._call_begin_hook(mode)
      File "/home/stweil/src/github/qurator-spk/train-calamari-gt4histocr/venv/lib/python3.6/site-packages/tensorflow/python/keras/callbacks.py", line 262, in _call_begin_hook
        self.on_train_begin()
      File "/home/stweil/src/github/qurator-spk/train-calamari-gt4histocr/venv/lib/python3.6/site-packages/tensorflow/python/keras/callbacks.py", line 378, in on_train_begin
        callback.on_train_begin(logs)
      File "/home/stweil/src/github/qurator-spk/train-calamari-gt4histocr/venv/lib/python3.6/site-packages/calamari_ocr/ocr/backends/tensorflow_backend/callbacks/earlystopping.py", line 67, in on_train_begin
        version=self.checkpoint_params.early_stopping_best_model_prefix,
      File "/home/stweil/src/github/qurator-spk/train-calamari-gt4histocr/venv/lib/python3.6/site-packages/calamari_ocr/ocr/backends/tensorflow_backend/callbacks/earlystopping.py", line 85, in make_checkpoint
        self.model.save(checkpoint_path + '.h5', overwrite=True)
      File "/home/stweil/src/github/qurator-spk/train-calamari-gt4histocr/venv/lib/python3.6/site-packages/tensorflow/python/keras/engine/network.py", line 1211, in save
        saving.save_model(self, filepath, overwrite, include_optimizer, save_format)
      File "/home/stweil/src/github/qurator-spk/train-calamari-gt4histocr/venv/lib/python3.6/site-packages/tensorflow/python/keras/saving/save.py", line 113, in save_model
        model, filepath, overwrite, include_optimizer)
      File "/home/stweil/src/github/qurator-spk/train-calamari-gt4histocr/venv/lib/python3.6/site-packages/tensorflow/python/keras/saving/hdf5_format.py", line 101, in save_model_to_hdf5
        default=serialization.get_json_type).encode('utf8')
      File "/usr/lib/python3.6/json/__init__.py", line 238, in dumps
        **kw).encode(obj)
      File "/usr/lib/python3.6/json/encoder.py", line 199, in encode
        chunks = self.iterencode(o, _one_shot=True)
      File "/usr/lib/python3.6/json/encoder.py", line 257, in iterencode
        return _iterencode(o, 0)
      File "/home/stweil/src/github/qurator-spk/train-calamari-gt4histocr/venv/lib/python3.6/site-packages/tensorflow/python/util/serialization.py", line 69, in get_json_type
        raise TypeError('Not JSON Serializable:', obj)
    TypeError: ('Not JSON Serializable:', b'\n\x08floordiv\x12\x08FloorDiv\x1a\x15input_sequence_length\x1a\nfloordiv/y*\x07\n\x01T\x12\x020\x01')

I opened this issue here to discuss it with other interested users and to collect information or potential solutions.

**Comment by stweil** [comment]
The same error also occurs on a normal PC with these Python packages:

    Package              Version
    -------------------- -------
    absl-py              0.8.1  
    astor                0.8.0  
    calamari-ocr         1.0.1  
    cycler               0.10.0 
    decorator            4.4.1  
    edit-distance        1.0.4  
    gast                 0.3.2  
    google-pasta         0.1.8  
    grpcio               1.25.0 
    h5py                 2.10.0 
    imageio              2.6.1  
    Keras-Applications   1.0.8  
    Keras-Preprocessing  1.1.0  
    kiwisolver           1.1.0  
    lxml                 4.4.2  
    Markdown             3.1.1  
    matplotlib           3.1.2  
    networkx             2.4    
    numpy                1.17.4 
    Pillow               6.2.1  
    pip                  18.1   
    pkg-resources        0.0.0  
    prettytable          0.7.2  
    protobuf             3.11.1 
    pyparsing            2.4.5  
    python-bidi          0.4.2  
    python-dateutil      2.8.1  
    PyWavelets           1.1.1  
    scikit-image         0.16.2 
    scipy                1.3.3  
    setuptools           42.0.2 
    six                  1.13.0 
    tensorboard          1.14.0 
    tensorflow           1.14.0 
    tensorflow-estimator 1.14.0 
    termcolor            1.1.0  
    tqdm                 4.40.1 
    Werkzeug             0.16.0 
    wheel                0.33.6 
    wrapt                1.11.2 
    XlsxWriter           1.2.6

**Comment by ChWick** [comment]
Calamari needs Tensorflow v2 from now on, try to upgrade it `pip install -U tensorflow`

**Comment by stweil** [comment]
I had tried with tensorflow 2.0.0b1, too, but got the same error. There is no newer version for Python 3.7 / Linux / amd64 on PyPI.

**Comment by amitdo** [comment]
https://pypi.org/project/tensorflow-gpu/#history
https://pypi.org/project/tensorflow-gpu/#files

**Comment by stweil** [comment]
@ChWick, I searched for the reason why I could not get newer versions of `tensorflow`. The reason was quite simple: I first had to get the latest `pip`. So `pip install -U pip && pip install -U tensorflow` got a working environment. Thank you.

See https://www.python.org/dev/peps/pep-0571/. Support for `manylinux2010` (which is responsible for the package versions which I get offered) was added in pip 19, so that is the minimum version required.
